test/src/unit-cppapi-string-dims.cc

```diff
@@ -35,6 +35,90 @@
 
 using namespace tiledb;
 
+static bool active_diags = false;
```

Comment:
@Shelnutt2 Looks like David wants to put diagnostic information for the test behind a boolean value... Is there any way we could do this in a more spdlog proper way?

Reply:
Is there any fine-grained approach to using that?
As the usage grows, seems the amount of information presented could potentially bury output from any one area making it difficult to see the particular information desired?
And, should the items defined in common/logger.h (LOG_TRACE(), LOG_DEBUG() etc.) be used rather than using global_logger().whatever() directly?

Reply:
Good point, I would vote for using the macros... ie: LOG_TRACE. This was we give ourselves the possibility to change all logging later.

Also maybe debug should be reserved for mostly non test code and we can use trace in test code. That way we can somewhat control the amount of debug spew. And in the test code case, when we enable trace level, we usually most likely will be running one test case so that should reduce the output. Thoughts @Shelnutt2 ?

Reply:
@KiterLuc `LOG_X` aren't macros, they are just wrapper functions calling the global logger. The global logger is our own logger class so we are abstracted from spdlog directly. Feel free to use the functions or global logger directly.

I'm on board with `TRACE` level for testing.

Reply:
@Shelnutt2 @KiterLuc 
I'm assuming there is a desire for removal of the local 'active_diags' flag variable controlling the formatting/generation of output.
Removing the flag and using the simple approach of changing std::cout to a std::stringstream variable and LOG_TRACE()ing that variable will mean the formatting code is always active.
Does this not matter?
Is there another simple approach I've not thought of?

Reply:
@Shelnutt2 @KiterLuc 
? Appears spdlog has a get_level() which our wrapper does not expose, could be exposed via our wrapper and used to modify the module control flag, supporting runtime spec of tracing via the log settings, while avoiding the runtime overhead of output formatting if tracing not requested... ?

Reply:
Search the code for LOG_STATUS. I think that since there's a lot of examples of using + to concatenate error strings which we should just do the same. And personally, I wouldn't worry too much about the extra processing in this specific case. But yes for cases where we might log a lot, it would be a great idea to be mindful of processing time.

Reply:
In the code are already a number of places using stringstream just as I intend, so I think I'll stick with that approach and avoid having to intrusively modify the formatting code already in place.

Reply:
Sure although the code would be more compact that way.

Reply:
Assuming you mean compactness of source code (is that what you meant?), 
I am doubtful given presence of non-string entities, 
think overall its prob. about even, maybe a loss, but open to being educated.
(Are there available implicit conversions to string I'm not aware of that could shorten the string versions?)
One comparison using what I am aware of, between one shorter reality in code and another longer reality,
stringstream version has more compact source code...
```
      int tot_result_num = -1;
      int othernum = 0;
      //stringstream...
      std::stringstream ssm; ssm << "tot_result_num " << tot_result_num << " othernum " << othernum << "\n"; LOG_TRACE(ssm.str());
      //various string alternates...
      std::string s1 = "tot_result_num " + tiledb::sm::utils::parse::to_str(tot_result_num) + " othernum " + tiledb::sm::utils::parse::to_str(othernum) + "\n"; LOG_TRACE(s1);
      std::string s2 = "tot_result_num " + std::to_string(tot_result_num) + " othernum " + std::to_string(othernum) + "\n"; LOG_TRACE(s2);
      std::stringstream sst1, sst2; sst1 << tot_result_num; sst2 << othernum; std::string s3 = "tot_result_num " + sst1.str() + " othernum " + sst2.str() + "\n"; LOG_TRACE(s3);
```
In statements where all entities are of string type, then those particular satements likely would be shorter (diff of "<<" vs "+" * n_uses_of_operator.)
Regarding compactness elsewhere,
looking at output from LOG_TRACE(), wonder if it should be used in situations such as this,
as it rather overwhelms the meat of the output data in this module with the ->set_pattern() line prefix overhead tiledb specifies.

Reply:
You can put the + version directly as a parameter to LOG_TRACE and not have a variable declaration... That's what I meant by more compact. But again let's not hold merging on this.

If you think it's too much output there's also a compile time version of log_trace. Ie:
#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_INFO // All DEBUG/TRACE statements will be removed by the pre-processor
#include "spdlog/spdlog.h"
...
SPDLOG_DEBUG("debug message to default logger"); // removed at compile time
SPDLOG_LOGGER_TRACE(my_logger, "trace message"); // removed at compile time
SPDLOG_INFO("info message to default logger");   // included

Reply:
The output I was referring to is what is actually displayed...
raw
```
option 9 current read_status() COMPLETE, (initial) result_num 3, final expected_result_num 3, bufcnt 6
a       2       j       4
bb      2       kk      5
c       1       i       3
tot_result_num 3
```
vs rather obfuscating 'not really useful for this case' logging prefix tiledb sets via logger->set_pattern() 
```
[2021-08-19 11:49:50.146] [tiledb] [Process: 37232] [Thread: 52624] [trace] option 9 current read_status() COMPLETE, (initial) result_num 3, final expected_result_num 3, bufcnt 6
[2021-08-19 11:49:50.147] [tiledb] [Process: 37232] [Thread: 52624] [trace] a   2       j       4
[2021-08-19 11:49:50.147] [tiledb] [Process: 37232] [Thread: 52624] [trace] bb  2       kk      5
[2021-08-19 11:49:50.147] [tiledb] [Process: 37232] [Thread: 52624] [trace] c   1       i       3
[2021-08-19 11:49:50.147] [tiledb] [Process: 37232] [Thread: 52624] [trace] tot_result_num 3
```

Reply:
Oh I'm sorry I got you now, thanks for clarifying. I would be ok with it... @Shelnutt2 thoughts?

Reply:
I'm good with whatever you guys prefer. The logger idea was most to standard things under the `trace` level, but its not a hard requirement for me.